op-e2e/actions/user.go

```diff
@@ -0,0 +1,322 @@
+	require.NoError(t, err, "failed to get L1 nonce for account %s", s.address)
+	return nonce
+}
+
+func (s *BasicUser[B]) TxValue() *big.Int {
+	if s.txOpts.Value != nil {
+		return s.txOpts.Value
+	}
+	return big.NewInt(0)
+}
+
+// ActMakeTx makes a tx with the predetermined contents (see randomization and other actions)
+// and sends it to the tx pool
+func (s *BasicUser[B]) ActMakeTx(t Testing) {
+	gas, err := s.env.EthCl.EstimateGas(t.Ctx(), ethereum.CallMsg{
+		From:      s.address,
+		To:        s.txToAddr,
+		GasFeeCap: s.txOpts.GasFeeCap,
+		GasTipCap: s.txOpts.GasTipCap,
+		Value:     s.TxValue(),
+		Data:      s.txCallData,
+	})
+	require.NoError(t, err, "gas estimation should pass")
+	tx := types.MustSignNewTx(s.account, s.env.Signer, &types.DynamicFeeTx{
+		To:        s.txToAddr,
+		GasFeeCap: s.txOpts.GasFeeCap,
+		GasTipCap: s.txOpts.GasTipCap,
+		Value:     s.TxValue(),
+		ChainID:   s.env.Signer.ChainID(),
+		Nonce:     s.PendingNonce(t),
+		Gas:       gas,
+	})
+	err = s.env.EthCl.SendTransaction(t.Ctx(), tx)
+	require.NoError(t, err, "must send tx")
+	s.lastTxHash = tx.Hash()
+}
+
+func (s *BasicUser[B]) ActCheckReceiptStatusOfLastTx(success bool) func(t Testing) {
+	return func(t Testing) {
+		s.CheckReceipt(t, success, s.lastTxHash)
+	}
+}
+
+func (s *BasicUser[B]) CheckReceipt(t Testing, success bool, txHash common.Hash) *types.Receipt {
+	receipt, err := s.env.EthCl.TransactionReceipt(t.Ctx(), txHash)
+	if receipt != nil && err == nil {
+		expected := types.ReceiptStatusFailed
+		if success {
+			expected = types.ReceiptStatusSuccessful
+		}
+		require.Equal(t, expected, receipt.Status, "expected receipt status to match")
+		return receipt
+	} else if err != nil && !errors.Is(err, ethereum.NotFound) {
+		t.Fatalf("receipt for tx %s was not found", txHash)
+	} else {
+		t.Fatalf("receipt error: %v", err)
+	}
+	return nil
+}
+
+type L1User struct {
+	BasicUser[*L1Bindings]
+}
+
+type L2User struct {
+	BasicUser[*L2Bindings]
+}
+
+type CrossLayerUser struct {
+	L1 L1User
+	L2 L2User
+
+	lastL1DepositTxHash common.Hash
+}
+
+func NewCrossLayerUser(log log.Logger, priv *ecdsa.PrivateKey, rng *rand.Rand) *CrossLayerUser {
+	addr := crypto.PubkeyToAddress(priv.PublicKey)
+	return &CrossLayerUser{
+		L1: L1User{
+			BasicUser: BasicUser[*L1Bindings]{
+				log:     log,
+				rng:     rng,
+				account: priv,
+				address: addr,
+			},
+		},
+		L2: L2User{
+			BasicUser: BasicUser[*L2Bindings]{
+				log:     log,
+				rng:     rng,
+				account: priv,
+				address: addr,
+			},
+		},
+	}
+}
+
+func (s *CrossLayerUser) ActDeposit(t Testing) {
+	isCreation := false
+	toAddr := common.Address{}
+	if s.L2.txToAddr == nil {
+		isCreation = true
+	} else {
+		toAddr = *s.L2.txToAddr
+	}
+	depositTransferValue := s.L2.TxValue()
+	depositGas := s.L2.txOpts.GasLimit
+	if s.L2.txOpts.GasLimit == 0 {
+		// estimate gas used by deposit
+		gas, err := s.L2.env.EthCl.EstimateGas(t.Ctx(), ethereum.CallMsg{
```

Comment:
Perhaps optionally add a buffer, this is known to produce faulty gas estimation

Reply:
The action can run just in time before the block creation, so should never be a bad estimate (and I'm curious when that estimate doesn't work). Alternatively we can set the gaslimit to non-0 with a separate action, and implement any special estimation there.